File: ios/GameCheckout/GameCheckout/BorrowerProfileStore.swift
import Foundation

/// Remembers the last name/email used to borrow a game so the borrow form
/// can prefill itself next time. Not sensitive, so UserDefaults (not
/// Keychain) is fine here.
enum BorrowerProfileStore {
    private static let nameKey = "borrowerProfile.name"
    private static let emailKey = "borrowerProfile.email"

    static var name: String {
        UserDefaults.standard.string(forKey: nameKey) ?? ""
    }

    static var email: String {
        UserDefaults.standard.string(forKey: emailKey) ?? ""
    }

    static func save(name: String, email: String) {
        if name != Self.name {
            UserDefaults.standard.set(name, forKey: nameKey)
        }
        if email != Self.email {
            UserDefaults.standard.set(email, forKey: emailKey)
        }
    }
}

File: android/app/src/main/java/ws/chill/gamecheckout/profile/BorrowerProfileStore.kt
package ws.chill.gamecheckout.profile

import android.content.Context

/**
 * Port of `BorrowerProfileStore.swift`.
 *
 * Remembers the last name/email used to borrow a game so the borrow form can
 * prefill itself next time. Not sensitive, so `SharedPreferences` (not
 * [ws.chill.gamecheckout.auth.SecureStore]) is fine here.
 */
class BorrowerProfileStore(context: Context) : BorrowerProfile {

    private val preferences =
        context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    override val name: String
        get() = preferences.getString(KEY_NAME, "").orEmpty()

    override val email: String
        get() = preferences.getString(KEY_EMAIL, "").orEmpty()

    override fun save(name: String, email: String) {
        val previousName = this.name
        val previousEmail = this.email

        val editor = preferences.edit()
        if (name != previousName) editor.putString(KEY_NAME, name)
        if (email != previousEmail) editor.putString(KEY_EMAIL, email)
        editor.apply()
    }

    private companion object {
        const val PREFS_NAME = "borrower_profile"
        const val KEY_NAME = "borrowerProfile.name"
        const val KEY_EMAIL = "borrowerProfile.email"
    }
}
